### DISCUSSION TITLE
Why is the config loaded twice and get_appearance returns different values?

### DISCUSSION THREAD

**Discussion by chtenb** [wezterm/wezterm#5922] [Q&A] [answered]
### What Operating System(s) are you running on?

Windows

### WezTerm version

wezterm 20240520-135708-b8f94c47

### Ask your question!

Consider this example config:

```
local wez = require "wezterm"
wez.log_info("get_appearance():")
wez.log_info(wez.gui.get_appearance())
return {}
```

When starting wezterm with this config, the log output is:

```
15:22:48.268  INFO   logging > lua: get_appearance():
15:22:48.270  INFO   logging > lua: Light
15:22:48.521  INFO   logging > lua: get_appearance():
15:22:48.522  INFO   logging > lua: Dark
```

Why are these log lines passed twice, and why is the value of get_appearance() different each time? My system is in Dark mode.
(Background: I'm troubleshooting get_appearance issues in my config, and I want/need to understand this aspect in order to proceed with troubleshooting)

**Comment by wez** [comment, marked as answer]
The configuration can be loaded multiple times from different contexts in various situations; this is documented at the bottom of this section in the docs: https://wezfurlong.org/wezterm/config/files.html#configuration-files

re: why is the appearance different on different calls? A number of window/gui related things take some time to load in and resolve. It is important to note that the configuration is loaded before we create any windows. Once they resolve, the configuration is reloaded and re-applied.

**Reply by chtenb** [reply]
Thanks for the pointer and explanation. When I open the debug overlay using ctrl-L, the configuration gets reloaded too, but this time only once, and only "Light" gets printed. It seems to not resolve to the Dark mode properly.

This is what the stdout looks like when opening wezterm and pressing ctrl-L several times

```
> wezterm --config-file repro.lua                                                                                                                                                          -1073741510 08/14/2024 10:52:36 AM
10:52:52.699  INFO   logging > lua: get_appearance():
10:52:52.701  INFO   logging > lua: Light
10:52:52.954  INFO   logging > lua: get_appearance():
10:52:52.955  INFO   logging > lua: Dark
10:52:56.614  INFO   logging > lua: get_appearance():
10:52:56.616  INFO   logging > lua: Light
10:52:57.218  INFO   logging > lua: get_appearance():
10:52:57.220  INFO   logging > lua: Light
10:52:58.149  INFO   logging > lua: get_appearance():
10:52:58.150  INFO   logging > lua: Light
```

Based on what I understand from the docs and your explanation I would expect always to see a "Dark" log event after any "Light" log event, since my system is in Dark mode. Is this a bug, or is my understanding incorrect?

**Reply by wez** [reply]
Feels like a bug!

**Reply by romac** [reply]
FWIW I have been running into the very same issue, which sometimes  caused my color scheme to be wrongly set to light mode instead of dark mode. I have managed to work around it by overriding the color scheme in the `window-config-reloaded` hook, but it still sometimes causes some flickering when opening a new window:

```lua
local function is_dark()
	if wt.gui then
		return wt.gui.get_appearance():find("Dark")
	end
	return true
end

local function scheme_for_appearance(theme)
	if is_dark() then
		return theme.dark
	else
		return theme.light
	end
end

wt.on("window-config-reloaded", function(window, _pane)
	local overrides = window:get_config_overrides() or {}
	local scheme = scheme_for_appearance({ dark = 'Catppuccin Dark', light = 'Catpuccin Light'})
	if overrides.color_scheme ~= scheme then
		overrides.color_scheme = scheme
		window:set_config_overrides(overrides)
	end
end)
```

**Reply by chtenb** [reply]
Issue created: https://github.com/wez/wezterm/issues/5982
